### ISSUE TITLE
BLSTMP does not make use of dropout

### ISSUE THREAD

**Issue by ozancaglayan** [espnet/espnet#259]
In `src/nets/e2e_asr_attctc_th.py:2045`, a dropout is given to the LSTM layers with `num_layers=1`. The dropout in PyTorch RNN layers are only effective if `num_layers > 1` since they are applied after each layer except the last layer.
https://pytorch.org/docs/0.3.1/nn.html?highlight=lstm#torch.nn.LSTM

**Comment by sw005320** [comment]
Thanks for the note. It seems to be true, and we may need to perform dropout explicitly at somewhere in https://github.com/espnet/espnet/blob/master/src/nets/e2e_asr_attctc_th.py#L2062-L2076.

**Comment by jnishi** [comment]
F.Y.I

It is said that applying dropout to recurrent neural networks simply does not work well. There are several ideas to apply dropout to LSTM, e.g. cell update vector only (https://arxiv.org/abs/1603.05118)
or output of each gate (https://www.isca-speech.org/archive/Interspeech_2017/pdfs/0129.PDF)

**Comment by arendu** [comment]
> It is said that applying dropout to recurrent neural networks simply does not work well.
Is this for dropout in the recurrent weights? or dropout between consecutive rnn layers?

**Comment by jnishi** [comment]
> Is this for dropout in the recurrent weights? or dropout between consecutive rnn layers?

Recurrent weights are (cf. https://arxiv.org/pdf/1409.2329.pdf)
Simplest way to apply dropout to rnn is to apply only between layers.
